.cruft.json

```diff
@@ -8,8 +8,8 @@
       "author_email": "cloud-service@elixir-europe.org",
       "development_status": "1 - Planning",
       "short_description": "File handler utilizing TUS and MinIO with DRS-Filer integration.",
-      "project_name": "tus_storagehandler",
-      "project_slug": "tus_storagehandler",
+      "project_name": "cloud-storage-handler",
```

Comment:
![image](https://github.com/user-attachments/assets/68d70d9a-e1d0-4ce0-aed2-562440500a68)

It is not allowed to use this name.
cloud-storage-handler works.

Reply:
Hmm, okay. It's a bit weird, because for places where we need to avoid whitespaces (assuming that's the problem), we already have the slug. @JaeAeich: Can you comment on this? I think it would be useful to have a project name that allows whitespace, e.g., for the title page and whenever anything is user-facing.

Reply:
@JaeAeich

Also, apart from the project titile, there is an issue with the slug as well:
- GitHub projects, PyPI projects, CLI tools should all use dashes instead of underscores (e.g., `cloud-storage-handler`). And PyPI projects and CLI tools should be all lowercase (GitHub projects not necessarily, but I'm okay with enforcing that as well).
- The actual code package MUST NOT include dashes and should use underscores instead (`cloud_storage_handler`).

My proposal: We ask users for a project title (which can contain whitespace and capital letters), transform it (limitations: all ASCII, all lowercase, whitespace to dashes, special characters removed, cut any numbers from the beginning of the string) to suggest a project slug (which can be overridden by the user as long as it conforms to the limitations) and then we create the underscore version of the slug ourselves by simply replacing dashes with underscores.

Example:

```console
Project name (user input): Cloud Storage Handler
Project slug (auto-suggeted; user can override): cloud-storage-handler
Alternative project slug (Python packages & modules; auto-generated): cloud_storage_handler
```

Are there any cases where this scheme does not cover our needs?

Reply:
- `project name` is for the name of the repo, this doesn't really has to do anythin with the project or what ever the proj is about. This is for github mainly. Note `whitespaces` aren't allowed as they aren't in github either, i.e. this is in consideration with the contraints from github.
  -  Why not just add the whitespaces? well, then the logic to create these template becomes cumbersome as there is no direct way to convert different case style and we will have to write pre and post hooks, which I think is the point of deminishing returns, introducing these complexity would be acceptable when we are dealing with helm charts maybe. @uniqueg Thought?
 - `project slug` this has constraints from pypi and/or poetry.
   - why aren't console scripts with dashes instead underscore? because pypi doesn't let us have dashes, ie the actual name name of the software/server/cli/project is `project_slug`, hence thats how it should be called from command line. But ofc no one is stopping us (rather I encourage it for atleast a shorthand name) from introducing multiple multiple console script targeting the same script, ie maybe we should also have `csh = cloud_storage_handler.app:main`, this way someone doesn't have to type the whole thing.  

@uniqueg's Proposal:
Well as you see apart from very basic transformations there is no way to do it without python script. From my definition `project_name` is the name of the repository that holds `project_slug`, and this `project_slug` si whats being released to pypi, this `project_slug` is how we target this project from console.

Just let me know if I rambled in the wrong direction or missed something.

---

For the title constraint that you have mentioned, I am not really sure, maybe I am mistaken here, but I did create a [test pacakge](https://test.pypi.org/manage/project/python-cookiecutter/releases/), and it worked fine. Please let me know if I am wrong. Assuming changes you requested are essential then I saw we should halt this PR and let pratiksha work on her project, these changes aren;t going to affect here project either way and look at the cookiecutter as that is the source of this mistake.

Reply:
I was not suggesting that we should use the project title (with  possible whitespace) _instead_ of the slugs - just that there is a need for human readable title.

My suggestion was to only ask for one type of slug (with dashes; for GitHub, console script etc.) and then create the other type of slug (with underscores; for PyPI etc.) internally. If that is too much of a hassle (or we want to allow users to set entirely different names for GitHub projects and PyPI packages), then we could also have three values:
- `project_title` (human readable; no restrictions)
- `project_slug_dashes` (instead of `project_name`, which is ambiguous)
- `project_slug_underscores` (instead of `project_slug`, which is ambiguous)

For suggesting defaults for the slugs from the project title, we could do the following sanitizations:
- replace all whitespace with <kbd>-</kbd> (`project_slug_dashes`) or <kbd>_</kbd> (`project_slug_underscores`)
- replace all uppercase with lowercase characters
- remove all characters except `[a-z0-9-]` (`project_slug_dashes`) or `[a-z0-9_]` (`project_slug_underscores`)
- remove leading numbers

For example, if the user sets `Cloud Storage Handler` as the `project_title`, then the suggested defaults for `project_slug_dashes` and `project_slug_underscores` will be `cloud-storage-handler` and `cloud_storage_handler`, respectively. Nothing will change in how those last two are used (other than their names), but instead of using project name in the repo and docs titles etc., `project_title` will be used.

But I am agreeing on the last part: Even though this issue has a bearing on the name change, we shouldn't keep blocking @psankhe28 for this. In the worst case, we lose the ability to sync/update via Cruft. So unless it is something that you can implement very quickly, @JaeAeich, let's move ahead.

Reply:
Ahh sure, yeah I can implement it rather quickly but lets do it after tom's meet (just to be sure) by then @psankhe28 can go ahead and implement her services.

Reply:
Sounds good. I'll close this then - to be taken care of automatically once the Cookiecutter is updated.